### PR TITLE
fix(common): reflect input type in context

### LINKED ISSUE #31556: Type check directive template context
# 🚀 feature request

### Relevant Package

@angular/compiler

### Description

Template contexts appear not to be checked.

```
    <ng-container *ngIf="data(); let d">
      {{ d.a }}
      {{ d.c }}
    </ng-container>
```

There is zero checking for the type of `d`.

Example: https://github.com/pauldraper/angular-template-context I would like `ng build` to fail.

### Describe the solution you'd like

Template context types are type checked.

### Describe alternatives you've considered

Don't use template contexts.

I've also tried using Ivy but it seems broken.

```
ERROR in __ng_typecheck__.ts:5:22 - error TS2304: Cannot find name 'Partial'.

5 const _ctor1: (init: Partial<Pick<i1.NgIf, "ngIf" | "ngIfThen" | "ngIfElse">>) => i1.NgIf = (null!);
```

### LINKED ISSUE #33527: Ivy: fullTemplateTypeCheck errors with NgIterable
# 🐞 bug report

### Affected Package

The issue is caused by package @angular/compiler

### Is this a regression?

This is introduced by Ivy.

### Description

A component with a template looking like:

```html
<div *ngFor="let user of users | slice: 0:2 as total">
  {{ total.length }}: {{ user.name }}
</div>
```

`ng build --prod` with `fullTemplateTypeCheck: true` throws, as it thinks `length` is not a property of `total`.

## 🔬 Minimal Reproduction

https://github.com/cexbrayat/ng-iterable

## 🔥 Exception or Error
<pre><code>
ERROR in src/app/app.component.html:2:6 - error TS2339: Property 'length' does not exist on type 'NgIterable<any>'.
  Property 'length' does not exist on type 'Iterable<any>'.

2   {{ total.length }}: {{ user.name }}
       ~~~~~~~~~~~~~

  src/app/app.component.ts:5:16
    5   templateUrl: './app.component.html',
                     ~~~~~~~~~~~~~~~~~~~~~~
    Error occurs in the template of component AppComponent.
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 9.0.0-rc.0
Node: 12.12.0
OS: darwin x64
Angular: 9.0.0-rc.0
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.0-rc.0
@angular-devkit/build-angular     0.900.0-rc.0
@angular-devkit/build-optimizer   0.900.0-rc.0
@angular-devkit/build-webpack     0.900.0-rc.0
@angular-devkit/core              9.0.0-rc.0
@angular-devkit/schematics        9.0.0-rc.0
@ngtools/webpack                  9.0.0-rc.0
@schematics/angular               9.0.0-rc.0
@schematics/update                0.900.0-rc.0
rxjs                              6.5.3
typescript                        3.6.4
webpack                           4.41.2
</code></pre>

**Anything else relevant?**

This is using ng and CLI `9.0.0-rc.0` and `fTTC`, but *not* using `strictTemplates`.

### PR DESCRIPTION
**Pull request by crisbeto**
Fixes `NgForOf` not reflecting the type of its input in the `NgForOfContext`. Also fixes `NgIf` which was typed to `any`.

Fixes #33527.
Fixes #31556.

### CODE REVIEW

**Comment by IgorMinar** [comment]
[presubmit](https://test.corp.google.com/ui#id=OCL:282499489:BASE:282499508:1574750327811:3bf9d550)
[ivy presubmit](https://test.corp.google.com/ui#id=OCL:282499489:BASE:282499650:1574750420705:1637d262)

**Review by IgorMinar — CHANGES_REQUESTED** [review]
can you please split this into two commits and document each change and what is the user-facing fix/impact because when users read about this change in the changelog they will not have any idea about what this does, or mean to them.

I'm assuming that this is a fix for #33527. Can you please state that in the commit message?

Lastly, any chance we we can add some tests for this? I'm assuming that we currently don't have any good infra to test typechecking changes like this unless we write a ngtsc specific test. Is that right?

I've kicked off a presubmit for this change in the meantime to make sure that there are no surprises.

**Comment by IgorMinar** [comment]
also, this should target the 9.0.x branch as well because otherwise we won't be able to roll this out until v10.

File: packages/common/src/directives/ng_if.ts [hunk 1]
```diff
@@ -149,14 +149,14 @@ import {Directive, EmbeddedViewRef, Input, TemplateRef, ViewContainerRef, ɵstri
  * @publicApi
  */
 @Directive({selector: '[ngIf]'})
-export class NgIf {
-  private _context: NgIfContext = new NgIfContext();
-  private _thenTemplateRef: TemplateRef<NgIfContext>|null = null;
-  private _elseTemplateRef: TemplateRef<NgIfContext>|null = null;
-  private _thenViewRef: EmbeddedViewRef<NgIfContext>|null = null;
-  private _elseViewRef: EmbeddedViewRef<NgIfContext>|null = null;
+export class NgIf<T> {
```
Comment (JoostK) [inline packages/common/src/directives/ng_if.ts]: There's currently nothing that allows for inferring the type `T` from the condition variable, which would be the `ngIf` input. Then, to allow the inferred type `T` to flow into `NgIfContext<T>` we'd need a `ngTemplateContextGuard` similar to that of `NgForOf`.

File: packages/common/src/directives/ng_if.ts [hunk 2]
```diff
@@ -230,9 +230,9 @@ export class NgIf {
 /**
  * @publicApi
  */
-export class NgIfContext {
-  public $implicit: any = null;
-  public ngIf: any = null;
+export class NgIfContext<T> {
+  public $implicit: T|null = null;
```
Comment (JoostK) [inline packages/common/src/directives/ng_if.ts]: I am a bit concerned about those `| null`s here, as they will be included in the type that the user needs to deal with. I reckon this is done to allow for the initial assignment to `null`, however I think we should "cheat" and assign `null !` here instead. The `NgIfContext<T>` should not be exposed before its conditions have been set, so we should be okay.

**Review by JoostK — CHANGES_REQUESTED** [review]
Regarding tests, I think the most valuable way would still be ngtsc's template type checking tests. I feel this way because those tests exercise real Angular templates.

As an alternative, a test could mimic the code that the type checker generates, as to assert that type inference works as expected. Such tests however are most appropriate to be run using https://github.com/microsoft/dtslint, which we don't currently have anywhere AFAIK. Also, I would consider them brittle as there's no guarantee that they are accurate, hence my preference for actual ngtsc template type checking tests.

**Comment by crisbeto** [comment]
I've addressed the feedback. Can you take another look @JoostK @IgorMinar?

File: packages/common/src/directives/ng_if.ts [hunk 3]
```diff
@@ -149,14 +149,14 @@ import {Directive, EmbeddedViewRef, Input, TemplateRef, ViewContainerRef, ɵstri
...
-  private _thenTemplateRef: TemplateRef<NgIfContext>|null = null;
-  private _elseTemplateRef: TemplateRef<NgIfContext>|null = null;
-  private _thenViewRef: EmbeddedViewRef<NgIfContext>|null = null;
-  private _elseViewRef: EmbeddedViewRef<NgIfContext>|null = null;
+export class NgIf<T> {
+  private _context: NgIfContext<T> = new NgIfContext<T>();
+  private _thenTemplateRef: TemplateRef<NgIfContext<T>> = null !;
+  private _elseTemplateRef: TemplateRef<NgIfContext<T>> = null !;
+  private _thenViewRef: EmbeddedViewRef<NgIfContext<T>> = null !;
+  private _elseViewRef: EmbeddedViewRef<NgIfContext<T>> = null !;
```
Comment (JoostK) [inline packages/common/src/directives/ng_if.ts]: All these private fields can keep the `| null`, as it keeps the implementation safe against nulls.

**Review by JoostK — DISMISSED** [review]
I would love to see template type checker tests. If you need help with that, please let me know.

**Review by IgorMinar — APPROVED** [review]
fyi: g3 is looking good. @crisbeto can you please address Joost's feedback?

**Comment by crisbeto** [comment]
I have all the feedback addressed locally, but it uncovered some issues with the new signatures. I'll pair up with Joost tomorrow.

**Comment by crisbeto** [comment]
All the issues and feedback should addressed now. Thank you for the help with the tests @JoostK!

**Review by JoostK — COMMENTED** [review]
In the `NgIf` commit, could you include a fixes note for #31556?

**Comment by crisbeto** [comment]
Done.

**Review by alxhub — APPROVED** [review]

**Comment by Splaktar** [comment]
It seems like there should be a test for how this typing behaves with `*ngIf="something | async; let thing". In that case, the behavior here doesn't seem to align with one of the most common use cases of `*ngIf`.

**Pull request closed without merging**